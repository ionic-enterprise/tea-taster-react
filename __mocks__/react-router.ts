import { vi } from 'vitest';

const replace = vi.fn();
const useHistory = vi.fn().mockReturnValue({ replace });

export { useHistory };
