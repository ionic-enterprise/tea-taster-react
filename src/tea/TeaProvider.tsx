import { ReactNode, createContext, useContext, useEffect, useState } from 'react';
import { Tea } from '../models';
import { client } from '../api/backend-api';
import { Preferences } from '@capacitor/preferences';

type Props = { children?: ReactNode };
type RawData = Omit<Tea, 'image' | 'rating'>;

const transform = async (data: RawData): Promise<Tea> => {
  const images: Array<string> = ['green', 'black', 'herbal', 'oolong', 'dark', 'puer', 'white', 'yellow'];
  const image = `/assets/images/${images[data.id - 1]}.jpg`;
  const { value } = await Preferences.get({ key: `rating${data.id}` });
  return { ...data, image, rating: parseInt(value || '0', 10) };
};

const unpack = (data: RawData[]): Promise<Tea[]> => {
  return Promise.all(data.map((t) => transform(t)));
};

type Context = { teas: Tea[]; rate: (id: number, rating: number) => Promise<void> };
const TeaContext = createContext<Context | undefined>(undefined);
const TeaProvider = ({ children }: Props) => {
  const [teas, setTeas] = useState<Tea[]>([]);

  const loadTeas = async () => {
    const { data } = await client.get<Omit<Tea, 'image'>[]>('/tea-categories');
    const teas = await unpack(data);
    setTeas(teas);
  };

  const rate = async (id: number, rating: number): Promise<void> => {
    await Preferences.set({ key: `rating${id}`, value: rating.toString() });
    const idx = teas.findIndex((t) => t.id === id);
    teas[idx].rating = rating;
    setTeas(teas);
  };

  useEffect(() => {
    loadTeas();
  }, []);

  return <TeaContext.Provider value={{ teas, rate }}>{children}</TeaContext.Provider>;
};

export const useTea = () => {
  const context = useContext(TeaContext);
  if (context === undefined) throw new Error('useTea must be used within TeaProvider');
  return context;
};

export default TeaProvider;
