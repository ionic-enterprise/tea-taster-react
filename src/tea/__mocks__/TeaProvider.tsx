import { vi } from 'vitest';

const rate = vi.fn().mockResolvedValue(undefined);
export const useTea = vi.fn(() => ({ teas: [], rate }));
