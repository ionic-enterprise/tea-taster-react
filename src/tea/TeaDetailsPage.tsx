import { IonBackButton, IonButtons, IonContent, IonHeader, IonImg, IonPage, IonTitle, IonToolbar } from '@ionic/react';
import { useTea } from './TeaProvider';
import { useParams } from 'react-router';
import { Tea } from '../models';
import { Rating } from '../shared/Rating';
import { useState } from 'react';

const TeaDetailsPage: React.FC = () => {
  const { id } = useParams<{ id: string }>();
  const { teas } = useTea();
  const tea: Tea | undefined = teas.find((t) => t.id === parseInt(id, 10));

  const [rating, setRating] = useState<number>(2);

  return (
    <IonPage>
      <IonHeader>
        <IonToolbar>
          <IonButtons slot="start">
            <IonBackButton defaultHref="/tea" />
          </IonButtons>
          <IonTitle>Tea Details</IonTitle>
        </IonToolbar>
      </IonHeader>
      <IonContent className="ion-padding">
        {tea && (
          <div>
            <div className="ion-justify-content-center" style={{ display: 'flex' }}>
              <IonImg src={tea.image} style={{ maxWidth: '75%', maxHeight: '512px' }} />
            </div>
            <h1 data-testid="name">{tea.name}</h1>
            <p data-testid="description">{tea.description}</p>
            <Rating rating={rating} onRatingChange={(r) => setRating(r)} />
          </div>
        )}
      </IonContent>
    </IonPage>
  );
};
export default TeaDetailsPage;
