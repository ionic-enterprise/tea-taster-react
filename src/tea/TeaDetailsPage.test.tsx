import { Mock, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { useTea } from './TeaProvider';
import { useParams } from 'react-router';
import TeaDetailsPage from './TeaDetailsPage';

vi.mock('../tea/TeaProvider');
vi.mock('react-router');

describe('<TeaDetailsPage />', () => {
  beforeEach(() => {
    (useParams as Mock).mockReturnValue({ id: '3' });
    (useTea as Mock).mockReturnValue({
      teas: [
        {
          id: 1,
          name: 'Green',
          image: '/assets/images/green.jpg',
          description: 'Green tea description.',
        },
        {
          id: 2,
          name: 'Black',
          image: '/assets/images/black.jpg',
          description: 'Black tea description.',
        },
        {
          id: 3,
          name: 'Herbal',
          image: '/assets/images/herbal.jpg',
          description: 'Herbal infusions description.',
        },
        {
          id: 4,
          name: 'Oolong',
          image: '/assets/images/oolong.jpg',
          description: 'Oolong tea description.',
        },
      ],
    });
    vi.clearAllMocks();
  });

  it('renders consistently', () => {
    const { asFragment } = render(<TeaDetailsPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the title', () => {
    const { baseElement } = render(<TeaDetailsPage />);
    const titleELements = baseElement.querySelectorAll('ion-title');
    expect(titleELements).toHaveLength(1);
  });

  it('renders the tea name', () => {
    render(<TeaDetailsPage />);
    expect(screen.getByTestId('name')).toHaveTextContent('Herbal');
  });

  it('renders the tea description', () => {
    render(<TeaDetailsPage />);
    expect(screen.getByTestId('description')).toHaveTextContent('Herbal infusions description.');
  });
});
