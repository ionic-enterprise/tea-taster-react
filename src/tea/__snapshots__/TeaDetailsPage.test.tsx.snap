// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TeaDetailsPage /> > renders consistently 1`] = `
<DocumentFragment>
  <div
    class="ion-page"
  >
    <ion-header>
      <!---->
      <ion-toolbar>
        <ion-buttons
          slot="start"
        >
          <ion-back-button
            default-href="/tea"
          />
        </ion-buttons>
        <ion-title>
          Tea Details
        </ion-title>
      </ion-toolbar>
    </ion-header>
    <ion-content
      class="ion-padding"
    >
      <div>
        <div
          class="ion-justify-content-center"
          style="display: flex;"
        >
          <ion-img
            src="/assets/images/herbal.jpg"
            style="max-width: 75%; max-height: 512px;"
          />
        </div>
        <h1
          data-testid="name"
        >
          Herbal
        </h1>
        <p
          data-testid="description"
        >
          Herbal infusions description.
        </p>
      </div>
    </ion-content>
  </div>
</DocumentFragment>
`;
